### PR TITLE
ufal/be-collection2item-table-is-not-migrated

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  6 |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |
## Problem description
I have a big problem with parsing `Content-Type: 'text-uri'`, because it cannot recognize new line string `\n` passed from Python. I wanted to use Vanilla endpoint `"/api/core/items/{ITEM_UUI}/mappedCollections/"` at first, but after a few hours of failing I decided to create a new one `/api/clarin/import/item/{ITEM_UUID}/mappedCollections`.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java [hunk 1]
```diff
@@ -366,6 +373,75 @@ public ItemRest importItem(HttpServletRequest request) throws SQLException, Auth
...
+            throw new RuntimeException("Context is null - cannot import item's mapped collections.");
+        }
+
+        // Load List of collection self links
+        List<String> requestAsStringList = new ArrayList<>();
+        JSONParser parser = new JSONParser();
+        try {
+            Object obj = parser.parse(new InputStreamReader(request.getInputStream()));
+
+            for (Object entity : (JSONArray) obj) {
```
Comment (Paurikova2) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:404]: maybe I will add some control if obj is JSONArray

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java [hunk 2]
```diff
@@ -366,6 +373,75 @@ public ItemRest importItem(HttpServletRequest request) throws SQLException, Auth
...
+    @RequestMapping(method = RequestMethod.POST, value = "/item" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID +
+            "/mappedCollections")
+    public void importItemCollections(@PathVariable UUID uuid, HttpServletRequest request) throws SQLException,
+            AuthorizeException, IOException {
+        Context context = ContextUtil.obtainContext(request);
+        if (Objects.isNull(context)) {
+            throw new RuntimeException("Context is null - cannot import item's mapped collections.");
+        }
+
+        // Load List of collection self links
```
Comment (Paurikova2) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:398]: do you mean single collection? If not, use collections

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java [hunk 3]
```diff
@@ -366,6 +373,75 @@ public ItemRest importItem(HttpServletRequest request) throws SQLException, Auth
...
+        } catch (Exception e) {
+            throw new RuntimeException("Cannot import item's mapped collections because parsing of the request JSON" +
+                    "throws this error: " + e.getMessage());
+        }
+
+        // Find Collections following its self link
+        List<DSpaceObject> listDsoFoundInRequest
+                = utils.constructDSpaceObjectList(context, requestAsStringList);
+
+        if (listDsoFoundInRequest.size() < 1) {
```
Comment (Paurikova2) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:417]: maybe use const for 1, but as you want

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java [hunk 4]
```diff
@@ -366,6 +373,75 @@ public ItemRest importItem(HttpServletRequest request) throws SQLException, Auth
...
+            throw new UnprocessableEntityException("Not a valid collection uuid.");
+        }
+
+        for (DSpaceObject dso : listDsoFoundInRequest) {
+
+            Item item = itemService.find(context, uuid);
+            if (dso != null && dso.getType() == COLLECTION && item != null) {
```
Comment (Paurikova2) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:424]: use Object.isNull() and equal

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java [hunk 5]
```diff
@@ -391,4 +467,12 @@ private Integer getIntegerFromString(String value) {
...
     }
-}
\ No newline at end of file
+
+    private boolean checkIfItemIsTemplate(Item item) {
+        return item.getTemplateItemOf() != null;
+    }
+
+    private boolean checkIfOwningCollection(Item item, UUID collectionID) {
+        return item.getOwningCollection().getID().equals(collectionID);
```
Comment (Paurikova2) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:476]: are you sure that item has collection?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java [hunk 6]
```diff
@@ -470,4 +472,53 @@ public void testImportAuthorityAndConfidenceInMetadata() throws Exception {
...
+                .withIssueDate("2017-10-17")
+                .withAuthor("Smith, Donald").withAuthor("Doe, John")
+                .withSubject("ExtraEntry")
+                .build();
+        context.restoreAuthSystemState();
+
+        // It is owning collection
+        String col1SelfLink = "https://localhost:8080/spring-rest/api/core/collections/" + col1.getID();
+        String col2SelfLink = "https://localhost:8080/spring-rest/api/core/collections/" + col2.getID();
+        String col3SelfLink = "https://localhost:8080/spring-rest/api/core/collections/" + col3.getID();
```
Comment (Paurikova2) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java:503]: use const for common parts of links

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java [hunk 7]
```diff
@@ -470,4 +472,53 @@ public void testImportAuthorityAndConfidenceInMetadata() throws Exception {
...
+        List<String> collectionSelfLinksList = new ArrayList<>();
+        collectionSelfLinksList.add(col1SelfLink);
+        collectionSelfLinksList.add(col2SelfLink);
+        collectionSelfLinksList.add(col3SelfLink);
+
+        ObjectMapper mapper = new ObjectMapper();
+        String token = getAuthToken(admin.getEmail(), password);
+
+        getClient(token).perform(post("/api/clarin/import/item/" +
```
Comment (Paurikova2) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java:512]: use const for link

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java [hunk 8]
```diff
@@ -470,4 +472,53 @@ public void testImportAuthorityAndConfidenceInMetadata() throws Exception {
...
+                                publicItem1.getID() + "/mappedCollections")
+                                .content(mapper.writeValueAsBytes(collectionSelfLinksList))
+                                .contentType(org.springframework.http.MediaType.APPLICATION_JSON))
+                        .andExpect(status().isOk());
+
+        Item updatedItem = itemService.find(context, publicItem1.getID());
+        assertEquals(updatedItem.getCollections().size(), 3);
```
Comment (Paurikova2) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java:519]: maybe get collections from updatedItem as var

**Review by Paurikova2 — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java (hunk 1 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:404]: check added

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java (hunk 3 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:417]: Used `CollectionUtils.isEmpty(listDsoFoundInRequest)` instead

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java (hunk 4 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:424]: That code is copied, I do not want to touch it

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java (hunk 5 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClarinItemImportController.java:476]: Added a check

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java (hunk 6 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java:503]: Done

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java (hunk 7 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java:512]: I do not used to use this pattern

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java (hunk 8 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/ClarinItemImportControllerIT.java:519]: It is not necessary

**Pull request merged**